File: May-LeetCoding-Challenge/10-Find-The-Town-Judge/Find-The-Town-Judge.scala
object Solution {
  def findJudge(N: Int, trust: Array[Array[Int]]): Int = {
    // number of people who trust same people
    val beTrusteds = Array.fill[Int](N)(0)
    // people who trust other or others
    val doTrusts = Array.fill[Boolean](N)(false)

    for (t <- trust) {
      // doTrust "trust" beTrusted
      val beTrusted = t(1)
      val doTrust = t(0)
      beTrusteds(beTrusted - 1) += 1
      doTrusts(doTrust - 1) = true
    }

    doTrusts
      .zip(beTrusteds)
      .zipWithIndex
      .find {
        case ((doTrust: Boolean, beTrusted: Int), _: Int) => !doTrust && beTrusted == N - 1
      }
      .map(_._2 + 1)
      .getOrElse(-1)
  }
}

File: May-LeetCoding-Challenge/10-Find-The-Town-Judge/Find-The-Town-Judge.cpp
class Solution {
public:
    int findJudge(int N, vector<vector<int>>& trust) {
        vector<int> deg_in(N+1, 0);
        vector<int> deg_out(N+1, 0);
        for (auto &v: trust){
            deg_out[v[0]]++;
            deg_in[v[1]]++;       
        }
        
        int ans = -1;
        for (int i = 1; i <= N; i++)
            if (deg_in[i] == N-1 and deg_out[i] == 0){
                if (ans == -1) ans = i;
                else return -1;
            }  
        return ans;
    }
};